File: js/temp-agent-trace.js
/**
 * Temporary script to test OpenAI agents with Braintrust tracing
 * 
 * Run with: node temp-agent-trace.js
 * 
 * Requirements:
 * - OPENAI_API_KEY environment variable
 * - BRAINTRUST_API_KEY environment variable (optional, can login via browser)
 */

async function main() {
  try {
    // Import the built modules
    const braintrust = await import('./dist/index.js');

    const { Agent, run, tool, addTraceProcessor, setTracingDisabled } = await import('@openai/agents');
    const { z } = await import('zod');

    console.log('Setting up Braintrust logger...');
    
    // Initialize Braintrust logging
    const logger = braintrust.initLogger({
      projectId: "37a8a4f4-4a34-4cdc-9b87-91ec0ebe9a97",
    });

    // Enhanced processor that logs raw data AND sends to Braintrust
    class EnhancedTracingProcessor extends braintrust.BraintrustTracingProcessor {
      onTraceStart(trace) {
        // console.log('\n=== RAW TRACE START ===');
        // console.log(JSON.stringify(trace, null, 2));
        return super.onTraceStart(trace);
      }
      
      onTraceEnd(trace) {
        // console.log('\n=== RAW TRACE END ===');
        // console.log(JSON.stringify(trace, null, 2));
        return super.onTraceEnd(trace);
      }
      
      onSpanStart(span) {
        // console.log('\n=== RAW SPAN START ===');
        // console.log(JSON.stringify(span, null, 2));
        return super.onSpanStart(span);
      }
      
      onSpanEnd(span) {
        // console.log('\n=== RAW SPAN END ===');
        // console.log(JSON.stringify(span, null, 2));
        return super.onSpanEnd(span);
      }
    }

    // Set up enhanced tracing (logs raw data + sends to Braintrust)
    const processor = new EnhancedTracingProcessor(logger);
    setTracingDisabled(false);
    addTraceProcessor(processor);

    console.log('Creating tools...');

    // Create a weather tool using the proper tool() helper
    const getWeatherTool = tool({
      name: "get_weather",
      description: "Get the current weather for a city",
      parameters: z.object({ city: z.string() }),
      execute: async (input) => {
        console.log(`🌤️  Getting weather for ${input.city}...`);
        return `The weather in ${input.city} is sunny with temperature 72°F and light winds.`;
      },
    });

    // Create a calculator tool for more complex interactions
    const calculatorTool = tool({
      name: "calculator",
      description: "Perform basic math calculations",
      parameters: z.object({ 
        operation: z.string().describe("The math operation to perform, e.g., '2 + 2' or '10 * 5'")
      }),
      execute: async (input) => {
        console.log(`🧮 Calculating: ${input.operation}`);
        try {
          // Simple eval for basic math (don't do this in production!)
          const result = eval(input.operation.replace(/[^0-9+\-*/().\s]/g, ''));
          return `The result of ${input.operation} is ${result}`;
        } catch (error) {
          return `Sorry, I couldn't calculate that. Please use basic math operations like +, -, *, /`;
        }
      },
    });

    console.log('Creating agent with tools...');

    // Create agent with tools
    const agent = new Agent({
      name: "weather-calc-agent",
      model: "gpt-4o-mini",
      instructions: `You are a helpful assistant that can get weather information and do calculations. 
        Use the get_weather tool when asked about weather in any city.
        Use the calculator tool when asked to do math.
        Be friendly and helpful!`,
      tools: [getWeatherTool, calculatorTool],
    });

    console.log('Running agent with tool calls...');

    // Run the agent with a prompt that will trigger tool usage
    const result = await run(agent, "What's the weather in San Francisco? Also, what's 15 * 24?");
    console.log(result.finalOutput);
    
    // Clean up
    await processor.shutdown();
    await logger.flush();
    
  } catch (error) {
    console.error('❌ Error:', error.message);  
    
    if (error.message.includes('@openai/agents')) {
      console.log('\n💡 Install @openai/agents first:');
      console.log('   npm install @openai/agents zod');
    }
    
    if (error.message.includes('OPENAI_API_KEY')) {
      console.log('\n💡 Set your OpenAI API key:');
      console.log('   export OPENAI_API_KEY=your_key_here');
    }
    
    process.exit(1);
  }
}

if (require.main === module) {
  main().catch(console.error);
}

File: py/temp_agent_trace.py
"""
Temporary script to test OpenAI agents with Braintrust tracing (Python version)

Run with: python temp_agent_trace.py

Requirements:
- OPENAI_API_KEY environment variable
- BRAINTRUST_API_KEY environment variable (optional, can login via browser)
- pip install openai-agents pydantic
"""

import asyncio
import os
from typing import Any


def main():
    try:
        # Import required modules
        import braintrust
        from braintrust.wrappers.openai import BraintrustTracingProcessor
        from agents import Agent, tracing, Runner
        from agents.tool import function_tool

        print("Setting up Braintrust logger...")

        # Initialize Braintrust logging
        logger = braintrust.init_logger(
            project_id="37a8a4f4-4a34-4cdc-9b87-91ec0ebe9a97"  # Replace with your actual project ID
        )

        # Enhanced processor that logs raw data AND sends to Braintrust
        class EnhancedTracingProcessor(BraintrustTracingProcessor):
            def on_trace_start(self, trace: tracing.Trace) -> None:
                # print('\n=== RAW TRACE START ===')
                # print(f"Trace: {trace.model_dump()}")
                return super().on_trace_start(trace)

            def on_trace_end(self, trace: tracing.Trace) -> None:
                # print('\n=== RAW TRACE END ===')
                # print(f"Trace: {trace.model_dump()}")
                return super().on_trace_end(trace)

            def on_span_start(self, span: tracing.Span[Any]) -> None:
                # print('\n=== RAW SPAN START ===')
                # print(f"Span: {span.model_dump()}")
                return super().on_span_start(span)

            def on_span_end(self, span: tracing.Span[Any]) -> None:
                # print('\n=== RAW SPAN END ===')
                # print(f"Span: {span.model_dump()}")
                return super().on_span_end(span)

        # Set up enhanced tracing (logs raw data + sends to Braintrust)
        processor = EnhancedTracingProcessor(logger)
        tracing.set_tracing_disabled(False)
        tracing.add_trace_processor(processor)

        print("Creating tools...")

        # Create a weather tool
        @function_tool
        def get_weather(city: str) -> str:
            """Get the current weather for a city"""
            print(f"🌤️  Getting weather for {city}...")
            return f"The weather in {city} is sunny with temperature 72°F and light winds."

        # Create a calculator tool for more complex interactions
        @function_tool
        def calculator(operation: str) -> str:
            """Perform basic math calculations. Operation should be like '2 + 2' or '10 * 5'"""
            print(f"🧮 Calculating: {operation}")
            try:
                # Simple eval for basic math (don't do this in production!)
                import re

                # Only allow basic math operations
                if re.match(r"^[\d+\-*/().\s]+$", operation):
                    result = eval(operation)
                    return f"The result of {operation} is {result}"
                else:
                    return "Sorry, I can only do basic math operations like +, -, *, /"
            except Exception as error:
                return f"Sorry, I couldn't calculate that. Please use basic math operations like +, -, *, /. Error: {error}"

        print("Creating agent with tools...")

        # Create agent with tools
        agent = Agent(
            name="weather-calc-agent",
            model="gpt-4o-mini",
            instructions="""You are a helpful assistant that can get weather information and do calculations. 
            Use the get_weather tool when asked about weather in any city.
            Use the calculator tool when asked to do math.
            Be friendly and helpful!""",
            tools=[get_weather, calculator],
        )

        print("Running agent with tool calls...")

        # Run the agent with a prompt that will trigger tool usage
        result = Runner.run_sync(agent, "What's the weather in San Francisco? Also, what's 15 * 24?")
        print(result.final_output_as(str))

        # Clean up
        processor.shutdown()
        logger.flush()

    except ImportError as e:
        if "agents" in str(e):
            print("❌ Error: Missing agents dependency")
            print("\n💡 Install agents first:")
            print("   pip install 'braintrust[openai-agents]'")
        else:
            print(f"❌ Import Error: {e}")
    except Exception as error:
        print(f"❌ Error: {error}")

        if "OPENAI_API_KEY" in str(error):
            print("\n💡 Set your OpenAI API key:")
            print("   export OPENAI_API_KEY=your_key_here")

        raise


if __name__ == "__main__":
    main()
